plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/intermediate/preserveorder/ModelOrderNodeComparator.java

```diff
@@ -0,0 +1,133 @@
+                if (p1Node != null && p1Node.equals(p2Node)) {
+                    // We are not allowed to look at the model order of the edges but we have to look at the actual
+                    // port ordering.
+                    for (LPort port : p1Node.getPorts()) {
+                        if (port.equals(p1SourcePort)) {
+                            return -1;
+                        } else if (port.equals(p2SourcePort)) {
+                            return 1;
+                        }
+                    }
+                    // Cannot happen, since both nodes have a connection to the previous layer.
+                    return Integer.compare(
+                            getModelOrderFromConnectedEdges(n1),
+                            getModelOrderFromConnectedEdges(n2));
+                }
+                
+                // Else the nodes are ordered by the nodes they connect to.
+                // One can disregard the model order here
+                // since the ordering in the previous layer does already reflect it.
+                for (LNode previousNode : previousLayer) {
+                    if (previousNode.equals(p1Node)) {
+                        return -1;
+                    } else if (previousNode.equals(p2Node)) {
+                        return 1;
+                    }
+                }
+            }
+            
+            // One node has no source port
+            if (!n1.hasProperty(InternalProperties.MODEL_ORDER) || !n2.hasProperty(InternalProperties.MODEL_ORDER)) {
+                return Integer.compare(
+                        getModelOrderFromConnectedEdges(n1),
+                        getModelOrderFromConnectedEdges(n2));
+            }
+            // Fall through case.
+            // Both nodes are not connected to the previous layer. Therefore, they must be normal nodes.
+            // The model order shall be used to order them.
+        }
+        // Order nodes by their order in the model.
+        return Integer.compare(
+                n1.getProperty(InternalProperties.MODEL_ORDER),
+                n2.getProperty(InternalProperties.MODEL_ORDER));
+    }
+    
+    /**
+     * The {@link InternalProperties#MODEL_ORDER} of the first incoming edge of a node.
+     * 
+     * @param n The node
+     * @return The model order of the first incoming edge of the given node.
+     * Returns Integer.MAX_VALUE if no such edge exists.
+     */
+    private int getModelOrderFromConnectedEdges(final LNode n) {
+        LPort sourcePort = n.getPorts().stream().filter(p -> !p.getIncomingEdges().isEmpty()).findFirst().orElse(null);
+        if (sourcePort != null) {
+            LEdge edge = sourcePort.getIncomingEdges().get(0);
+            if (edge != null) {
+                return edge.getProperty(InternalProperties.MODEL_ORDER);
+            }
+        }
+        // Set to -1 to sort nodes without a connection to the previous layer over dummy nodes.
```

Comment:
Is the `-1` still valid?

Reply:
Yes, if one returns -1 instead of Integer.MAX_VALUE nodes without a connection to a previous layer will be sorted above dummy nodes.